R/dependecies.R

```diff
@@ -0,0 +1,593 @@
+#' Get dependencies for all elements in lock
+get_ordered_dependencies <- function(lock) {
+   cat(crayon::blue(paste("Retrieving dependency info...")))
+   names(lock) <- vapply(lock, `[[`, character(1), "name")
+   deps <- get_dependencies_for_list(lock, lock, list())
+   cat("\n")
+   deps
+}
+
+#' Recursive function to take a list and lock and extract dependencies, sorting
+#' along the way using the combine_dependencies function.
+#' @param master_list list.  Packages to go through and parse dependencies from
+#' @param lock list.  Original list of locked packages
+#' @param previously_parsed_deps.  List of packages and their dependencies that
+#'   we have already parsed out of their respective description files
+get_dependencies_for_list <- function(master_list, lock, previously_parsed_deps) {
+  # Start off with the master list as our set of packages
+  current_dependencies <- master_list
+  for (i in seq_along(master_list)) {
+    package <- master_list[[i]]
+
+    # Create a new list for this package if it no dependencies have been parsed
+    # for any version of the package.
+    if(!package$name %in% names(previously_parsed_deps)) {
+      previously_parsed_deps[[package$name]] <- list()
+    }
+    if (!paste(package$version) %in% 
+      names(previously_parsed_deps[[package$name]])) {
+      dependency_output <- get_dependencies(
+       structure(package
+         , class = c(package$remote %||% "CRAN"
+           , class(package)))
+       , lock)
+      package <- dependency_output$package
+      single_package_dependencies <- dependency_output$dependencies
+
+      # Our package comes out of dependency search with a download_path
+      # attached.  We will use this for installation (assuming we keep this
+      # version of the package.
+      current_dependencies[[package$name]] <- package
+
+      # Store the dependencies and corresponding package object in our humongous
+      # previously_parsed_deps object using version as the second key level
+      previously_parsed_deps[[package$name]][[paste(package$version)]] <-
+        list(package = package, dependencies = single_package_dependencies)
+    } else {
+      single_package_dependencies <-
+        previously_parsed_deps[[package$name]][[
+          paste(package$version)]]$dependencies
```

Comment:
Ironically, a lot of this code could be much easier to read if more things were on one line.